### PR TITLE
[test] Testing the environment of the login nodes when accessed remotely over SSH

### PR DESCRIPTION
**Pull request by rsarm**
Closes https://github.com/eth-cscs/reframe/issues/510

### CODE REVIEW

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    def __init__(self):
+        super().__init__()
+        self.descr = 'Test the environment when accessed remotely over SSH'
+        self.valid_systems = ['daint:login', 'dom:login']
+        self.valid_prog_environs = ['PrgEnv-cray']
+        self.executable = "ssh"
+        self.executable_opts = ["%s" % self.current_system.name,
+                                "echo", "'$SCRATCH'"]
+        self.sanity_patterns = sn.assert_found(
+            r'/scratch/[\S+]', self.stdout)
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]: This should fit in a single line.

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+
+@rfm.simple_test
+class SSHLoginEnvCheck(rfm.RunOnlyRegressionTest):
+    def __init__(self):
+        super().__init__()
+        self.descr = 'Test the environment when accessed remotely over SSH'
+        self.valid_systems = ['daint:login', 'dom:login']
+        self.valid_prog_environs = ['PrgEnv-cray']
+        self.executable = "ssh"
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]: Please use single quotes.

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+        self.executable_opts = ["%s" % self.current_system.name,
+                                "echo", "'$SCRATCH'"]
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]: Single quotes.

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+@rfm.simple_test
+class SSHLoginEnvCheck(rfm.RunOnlyRegressionTest):
+    def __init__(self):
+        super().__init__()
+        self.descr = 'Test the environment when accessed remotely over SSH'
+        self.valid_systems = ['daint:login', 'dom:login']
+        self.valid_prog_environs = ['PrgEnv-cray']
+        self.executable = "ssh"
+        self.executable_opts = ["%s" % self.current_system.name,
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]:
```suggestion
        self.executable_opts = [self.current_system.name,
```

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+                                "echo", "'$SCRATCH'"]
+        self.sanity_patterns = sn.assert_found(
+            r'/scratch/[\S+]', self.stdout)
+        self.maintainers = ['RS', 'VK']
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]:
```suggestion
        self.maintainers = ['RS', 'LM']
```

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        self.tags = {'maintenance'}
```
Comment (teojgo) [inline cscs-checks/prgenv/ssh_environ_check.py]: Add also production so that we can catch if there is something problematic with the profiles.

**Comment by lucamar** [comment]
Thanks @rsarm: is it possible to check also the items proposed in #524 ?

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+@rfm.simple_test
+class SSHLoginEnvCheck(rfm.RunOnlyRegressionTest):
+    def __init__(self):
+        super().__init__()
+        self.descr = 'Test the environment when accessed remotely over SSH'
```
Comment (victorusu) [inline cscs-checks/prgenv/ssh_environ_check.py]: The test name and description are not very informative. I would vote to have this test parameterised, with the individual env variables that we want to check. I would extend the variables to `$SCRATCH`, `$XDG_DATA_DIRS`, `$XDG_RUNTIME_DIR`, `$MODULEPATH`, `$MODULESHOME` and finally on Cray we should have `CRAY_CPU_TARGET`, `CRAYPE_NETWORK_TARGET` and `PE_PRODUCT_LIST`.

This way, I would vote to have the test description to **check if %(env_var)s is defined when accessed remotely over SSH**

**Comment by vkarak** [comment]
@rsarm The problem you are seeing is due to the fact that a regression test is deep copied for each combination of system partition and programming environment, but the closure of the lambda function still refers to the original object, because functions are not deep copied. More specifically, when you create the lambda function for the `map()`, its closure contains `self`. At the time the lambda is defined `self` refers to the object created during loading of the test (by the `@simple_test` decorator). Before executing the test, ReFrame clones it, but the `self` of its closure still refers to the original object, which is not executed. As a result, `self._job` is `None`. Here is a small reproducer that showcases the problem:

```python
import copy


class C:
    def __init__(self):
        self.fn = lambda: self


if __name__ == '__main__':
    c = C()
    d = copy.deepcopy(c)
    print('c is bound to', c)
    print('d is bound to', d)
    print('c.fn() is bound to', c.fn())
    print('d.fn() is bound to', d.fn())
```

This will print the following, which demonstrates exactly the problem:

```
c is bound to <__main__.C object at 0x2b073e23e780>
d is bound to <__main__.C object at 0x2b073e267470>
c.fn() is bound to <__main__.C object at 0x2b073e23e780>
d.fn() is bound to <__main__.C object at 0x2b073e23e780>
```

A workaround to this bug is to avoid functions with closures. So you could rewrite your sanity function as follows:

```python
    def __init__(self):
        ...
        self.sanity_patterns = sn.all(
            sn.map(self.assert_envvar, ['CRAY_CPU_TARGET=haswell']))

    def assert_envvar(self, v):
        return sn.assert_found(v, self.stdout)
```

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+            'SCRATCH': r'/scratch/[\S+]',
+            'XDG_RUNTIME_DIR': r'/run/user/[\d+]'
+        }
+        self.executable = 'ssh'
+        echo_args = ' '.join('%s=$%s' % (i, i)  for i in reference.keys())
+        self.executable_opts = [self.current_system.name,
+                                'echo', "'%s'" % echo_args]
+        self.sanity_patterns = sn.all(sn.map(self.assert_envvar,
+                                      list(reference.keys()),
+                                      list(reference.values())))
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]: You can rewrite this in a simpler way:

```python
        self.sanity_patterns = sn.all(
            sn.map(self.assert_envvar, list(reference.items())))
```

See below how the `assert_envvar` is also simplified.

File: cscs-checks/prgenv/ssh_environ_check.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        self.maintainers = ['RS', 'LM']
+        self.tags = {'maintenance'}
+
+    def assert_envvar(self, var, value):
+        return sn.assert_found('='.join([var, value]), self.stdout)
```
Comment (vkarak) [inline cscs-checks/prgenv/ssh_environ_check.py]:
```python
    def assert_envvar(self, v):
        return sn.assert_found('='.join(v), self.stdout)
```

**Comment by vkarak** [comment]
@rsarm Can you revert your latest changes? I have fixed the PR myself yesterday, but forgot to merge it.

**Pull request merged**